In [1]:
import time
notebook_start_time = time.time()

import os
import sys

from datetime import datetime

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.preprocessing import LabelEncoder

pd.set_option('MAX_COLUMNS', 50)

In [2]:
matches = list(filter(lambda x: x[-3:]=='csv', os.listdir()))
print('Number of matches in the data : {0}'.format(len(matches)))

Number of matches in the data : 751


In [3]:
def delete_files(address):
    if address not in os.listdir():
        os.mkdir(address)
        return 0
    for file in os.listdir(address):
        os.remove(address+'/'+file)
    return 1 

In [4]:
hashlist = []

delete_files('balldata')

for match in matches:
    hashmap = {}
    with open(match) as matchdata, open('temp.csv', 'w') as out:
        for line in matchdata:
            balldata = line.split(',')
            if balldata[0]=="info":
                #print(balldata)
                if balldata[1] not in hashmap:
                    hashmap[balldata[1]] = balldata[2][:-1]
                else:
                    hashmap[balldata[1]+'2'] = balldata[2][:-1]
            elif balldata[0]=="ball":
                out.write(line)
        
        hashmap['match_id'] = match        
        hashlist.append(hashmap)
    os.rename('temp.csv', 'balldata/'+match)

In [5]:
hashmap

{'team': 'Royal Challengers Bangalore',
 'team2': 'Sunrisers Hyderabad',
 'gender': 'male',
 'season': '2016',
 'date': '2016/05/29',
 'competition': 'Indian Premier League',
 'venue': 'M Chinnaswamy Stadium',
 'city': 'Bangalore',
 'toss_winner': 'Sunrisers Hyderabad',
 'toss_decision': 'bat',
 'player_of_match': 'BCJ Cutting',
 'umpire': 'HDPK Dharmasena',
 'umpire2': 'BNJ Oxenford',
 'reserve_umpire': 'A Deshmukh',
 'tv_umpire': 'AK Chaudhary',
 'match_referee': 'RS Madugalle',
 'winner': 'Sunrisers Hyderabad',
 'winner_runs': '8',
 'match_id': '981019.csv'}

In [6]:
metadata = pd.DataFrame(hashlist)
meta = metadata.copy()

In [7]:
cols = ['ball', 'innings', 'over', 'team', 'striker', 'non-striker', 'bowler', 'runs', 'extras', 'wicket', 'out_batsmen']

os.chdir('balldata')

for match in os.listdir():
    try:
        df = pd.read_csv(match, header=None)
        df.columns = cols
        df.to_csv(match, index=False)
    except:
        print(match)
        os.remove(match)
        os.remove('../'+match)
        

os.chdir('..')

In [8]:
def convertDate(d):
    if type(d)==float:
        return np.nan
    else:
        return datetime.strptime(d, "%Y/%m/%d")

metadata['date'] = meta['date'].map(lambda x: convertDate(x))

In [9]:
metadata.sort_values(by="date", inplace=True)

In [10]:
metadata.head()

city            competition       date date2 eliminator gender  \
174   Bangalore  Indian Premier League 2008-04-18   NaN        NaN   male   
176       Delhi  Indian Premier League 2008-04-19   NaN        NaN   male   
175  Chandigarh  Indian Premier League 2008-04-19   NaN        NaN   male   
177      Mumbai  Indian Premier League 2008-04-20   NaN        NaN   male   
178     Kolkata  Indian Premier League 2008-04-20   NaN        NaN   male   

       match_id match_number      match_referee method neutralvenue outcome  \
174  335982.csv            1          J Srinath    NaN          NaN     NaN   
176  335984.csv            3       GR Viswanath    NaN          NaN     NaN   
175  335983.csv            2  S Venkataraghavan    NaN          NaN     NaN   
177  335985.csv            5          J Srinath    NaN          NaN     NaN   
178  335986.csv            4        FM Engineer    NaN          NaN     NaN   

    player_of_match reserve_umpire   season                         team  \
174     BB McCullum    VN Kulkarni  2007/08  Royal Challengers Bangalore   
176     MF Maharoof            NaN  2007/08             Delhi Daredevils   
175      MEK Hussey    MSS Ranawat  2007/08              Kings XI Punjab   
177      MV Boucher    SN Bandekar  2007/08               Mumbai Indians   
178       DJ Hussey        F Gomes  2007/08        Kolkata Knight Riders   

                           team2 toss_decision                  toss_winner  \
174        Kolkata Knight Riders         field  Royal Challengers Bangalore   
176             Rajasthan Royals           bat             Rajasthan Royals   
175          Chennai Super Kings           bat          Chennai Super Kings   
177  Royal Challengers Bangalore           bat               Mumbai Indians   
178              Deccan Chargers           bat              Deccan Chargers   

          tv_umpire     umpire         umpire2  \
174       AM Saheba  Asad Rauf     RE Koertzen   
176       IL Howell  Aleem Dar  GA Pratapkumar   
175       RB Tiffin  MR Benson      SL Shastri   
177  AV Jayaprakash   SJ Davis       DJ Harper   
178       Asad Rauf  BF Bowden     K Hariharan   

                                  venue                       winner  \
174               M Chinnaswamy Stadium        Kolkata Knight Riders   
176                    Feroz Shah Kotla             Delhi Daredevils   
175  "Punjab Cricket Association Stadiu          Chennai Super Kings   
177                    Wankhede Stadium  Royal Challengers Bangalore   
178                        Eden Gardens        Kolkata Knight Riders   

    winner_runs winner_wickets  
174         140            NaN  
176         NaN              9  
175          33            NaN  
177         NaN              5  
178         NaN              5

# Let's explore NaN values (missing values).

In [11]:
metadata.isnull().sum()

city                 0
competition          0
date                 0
date2              749
eliminator         742
gender               0
match_id             0
match_number        46
match_referee        0
method             732
neutralvenue       674
outcome            738
player_of_match      4
reserve_umpire      25
season               0
team                 0
team2                0
toss_decision        0
toss_winner          0
tv_umpire            2
umpire               1
umpire2              1
venue                0
winner              13
winner_runs        414
winner_wickets     350
dtype: int64

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
    <div><strong>There are 5 NaN values in several columns.</strong></div>
    <div><strong>Let's see if it is just coincidence(which I don't like) or do we have 5 empty rows.... </strong></div>
</div>

In [12]:
metadata.loc[(metadata.city.isnull())]

Empty DataFrame
Columns: [city, competition, date, date2, eliminator, gender, match_id, match_number, match_referee, method, neutralvenue, outcome, player_of_match, reserve_umpire, season, team, team2, toss_decision, toss_winner, tv_umpire, umpire, umpire2, venue, winner, winner_runs, winner_wickets]
Index: []

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">

<div><strong>files 1136561.csv-1136565.csv are empty</strong></div>

<div>Deleting these empty rows is a reasonable option now.</div>
</div>

In [13]:
metadata = metadata.loc[np.logical_not(metadata.city.isnull())]
meta = meta.loc[np.logical_not(meta.city.isnull())]
metadata.isnull().sum()

city                 0
competition          0
date                 0
date2              749
eliminator         742
gender               0
match_id             0
match_number        46
match_referee        0
method             732
neutralvenue       674
outcome            738
player_of_match      4
reserve_umpire      25
season               0
team                 0
team2                0
toss_decision        0
toss_winner          0
tv_umpire            2
umpire               1
umpire2              1
venue                0
winner              13
winner_runs        414
winner_wickets     350
dtype: int64

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
    <div>Yaay..!! now a whole lot of columns are clear with <strong>Zero</strong> missing values.</div>

<div>Don't get too excited already, we still have a lot of missing values to deal with.</div>

<div><strong>Date2</strong> has 749 missing values. Since, having a secondary date for a match is very rare it makes sense to have a lot of missing values.</div>

<h2> Data Processing </h2>
<ul>
    <li>As the matches with a not-NaN value for <strong>date2</strong> are played on date2, we are going to have an ultimate date variable <strong>"match_date"</strong> holding the date on which the match was held on.</li>
    <li><strong>Gender</strong> will be removed all the matches are of Men's category and hence <strong>Gender</strong> variable holds <strong>zero</strong> information.</li>
    <li><strong>Neutral Venue</strong> will also converted to boolean</li>
    <li>Umpires don't have any influence on the outcome of a game(Assuming they are fair). As of now, I am not going to consider Umpire data for further analysis</li>
    <li><strong>Winner</strong> and <strong>outcome</strong> go hand in hand. So, combining these two variables will make our data clear.</li>
</ul>

<div>After cleaning the data as mentioned above, we are going to have a look at <strong>match_number, method, player_of_match, eliminator</strong> columns.</div>
</div>

In [14]:
#merging date and date2
match_dates = []
for date1, date2 in zip(metadata.date.values, metadata.date2.values):
    if type(date2)==float:
        match_dates.append(date1)
    else:
        match_dates.append(date2)

metadata['match_date'] = match_dates
metadata.drop(['date', 'date2'], axis=1, inplace=True)

In [15]:
#removing umpire related variables
metadata.drop('gender', axis=1, inplace=True)
metadata['neutralvenue'] = meta['neutralvenue'].map(lambda x: 0 if type(x)==float else 1)
metadata.drop(['reserve_umpire', 'tv_umpire', 'umpire', 'umpire2', 'match_referee'], axis=1, inplace=True)

In [16]:
#merging outcome and winner attribute values.
outcomes=[]
for out, win in zip(metadata['outcome'], metadata['winner']):
    if type(win)==float:
        outcomes.append(out)
    else:
        outcomes.append(win)

metadata['winner'] = outcomes
metadata.drop('outcome', axis=1, inplace=True)

<h2 style="text-align: center; font-size:2.5em;"> match_number </h2>

In [17]:
metadata.loc[(metadata.match_number.isnull())].season.value_counts()

2018       4
2012       4
2009/10    4
2019       4
2016       4
2011       4
2013       4
2014       4
2017       4
2015       4
2009       3
2007/08    3
Name: season, dtype: int64

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
    <div>The variable <strong>match_number</strong> has several null values. But the null values are evenly divided among the seasons(3 in first two seasons and 4 in rest of them). </div>
    <div>Again, this cannot be coincidence(or null values have some meaning here). They probably represent knockout-stage matches.</div>
    <div>Lets see if our initial guess is true.</div>
</div>
    

In [18]:
metadata.loc[(metadata.match_number.isnull())]

city            competition eliminator     match_id  \
229         Mumbai  Indian Premier League        NaN   336038.csv   
230         Mumbai  Indian Premier League        NaN   336039.csv   
231         Mumbai  Indian Premier League        NaN   336040.csv   
286      Centurion  Indian Premier League        NaN   392237.csv   
287   Johannesburg  Indian Premier League        NaN   392238.csv   
288   Johannesburg  Indian Premier League        NaN   392239.csv   
345         Mumbai  Indian Premier League        NaN   419162.csv   
346         Mumbai  Indian Premier League        NaN   419163.csv   
347         Mumbai  Indian Premier League        NaN   419164.csv   
348         Mumbai  Indian Premier League        NaN   419165.csv   
418         Mumbai  Indian Premier League        NaN   501268.csv   
419         Mumbai  Indian Premier League        NaN   501269.csv   
420        Chennai  Indian Premier League        NaN   501270.csv   
421        Chennai  Indian Premier League        NaN   501271.csv   
492           Pune  Indian Premier League        NaN   548378.csv   
493      Bangalore  Indian Premier League        NaN   548379.csv   
494        Chennai  Indian Premier League        NaN   548380.csv   
495        Chennai  Indian Premier League        NaN   548381.csv   
568          Delhi  Indian Premier League        NaN   598070.csv   
569          Delhi  Indian Premier League        NaN   598071.csv   
570        Kolkata  Indian Premier League        NaN   598072.csv   
571        Kolkata  Indian Premier League        NaN   598073.csv   
628        Kolkata  Indian Premier League        NaN   734043.csv   
629         Mumbai  Indian Premier League        NaN   734045.csv   
630         Mumbai  Indian Premier League        NaN   734047.csv   
631      Bangalore  Indian Premier League        NaN   734049.csv   
687         Mumbai  Indian Premier League        NaN   829817.csv   
688           Pune  Indian Premier League        NaN   829819.csv   
689         Ranchi  Indian Premier League        NaN   829821.csv   
690        Kolkata  Indian Premier League        NaN   829823.csv   
747      Bangalore  Indian Premier League        NaN   981013.csv   
748          Delhi  Indian Premier League        NaN   981015.csv   
749          Delhi  Indian Premier League        NaN   981017.csv   
750      Bangalore  Indian Premier League        NaN   981019.csv   
55          Mumbai  Indian Premier League        NaN  1082647.csv   
56       Bangalore  Indian Premier League        NaN  1082648.csv   
57       Bangalore  Indian Premier League        NaN  1082649.csv   
58       Hyderabad  Indian Premier League        NaN  1082650.csv   
110         Mumbai  Indian Premier League        NaN  1136617.csv   
111        Kolkata  Indian Premier League        NaN  1136618.csv   
112        Kolkata  Indian Premier League        NaN  1136619.csv   
113         Mumbai  Indian Premier League        NaN  1136620.csv   
170        Chennai  Indian Premier League        NaN  1181764.csv   
171  Visakhapatnam  Indian Premier League        NaN  1181766.csv   
172  Visakhapatnam  Indian Premier League        NaN  1181767.csv   
173      Hyderabad  Indian Premier League        NaN  1181768.csv   

    match_number method  neutralvenue    player_of_match   season  \
229          NaN    NaN             0          SR Watson  2007/08   
230          NaN    NaN             0            M Ntini  2007/08   
231          NaN    NaN             0          YK Pathan  2007/08   
286          NaN    NaN             1       AC Gilchrist     2009   
287          NaN    NaN             1          MK Pandey     2009   
288          NaN    NaN             1           A Kumble     2009   
345          NaN    NaN             0         KA Pollard  2009/10   
346          NaN    NaN             0       DE Bollinger  2009/10   
347          NaN    NaN             0           A Kumble  2009/10   
348          NaN    NaN             0           SK Raina  2009/10   
418        

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
<ul>
<li>Indeed, all the matches with no match_number attribute correspond to knockout/playoff matches. IPL used to follow two semi-final format during the initial seasons. IPL officials later changed the structure of playoffs/knockouts favouring the top two teams that out-performed rest of the teams in the league stage of the game. Former(2 semi-final) format comprised of three matches after the league stage of IPL whereas the latter format resulted in 4 matches after the round-robin part of the league.</li>

<li>As the <strong>match_number</strong> is just a serial number we can remove <strong>match_number</strong> variable.</li>
<li>According to our exploration, <em><strong>null values in match_number represent knockout stage matches</strong></em>. We can retain this information in a new column before removing the match_number column.</li>
</ul>
</div>

In [19]:
# adding 'is_league_match' attribute to the data
metadata['is_league_match'] = meta['match_number'].isnull().astype(int)

#removing 'match_number' attribute from the data.
metadata.drop('match_number', axis=1, inplace=True)

<h2 style="text-align: center; font-size:2.5em;"> Eliminator </h2>

In [20]:
metadata.loc[np.logical_not(metadata.eliminator.isnull())]

city            competition                   eliminator  \
240  Cape Town  Indian Premier League             Rajasthan Royals   
304    Chennai  Indian Premier League              Kings XI Punjab   
502  Hyderabad  Indian Premier League          Sunrisers Hyderabad   
515  Bangalore  Indian Premier League  Royal Challengers Bangalore   
590  Abu Dhabi  Indian Premier League             Rajasthan Royals   
650  Ahmedabad  Indian Premier League              Kings XI Punjab   
33      Rajkot  Indian Premier League               Mumbai Indians   
123      Delhi  Indian Premier League               Delhi Capitals   
164     Mumbai  Indian Premier League               Mumbai Indians   

        match_id method  neutralvenue player_of_match   season  \
240   392190.csv    NaN             1       YK Pathan     2009   
304   419121.csv    NaN             0        J Theron  2009/10   
502   598004.csv    NaN             0       GH Vihari     2013   
515   598017.csv    NaN             0         V Kohli     2013   
590   729315.csv    NaN             1     JP Faulkner     2014   
650   829741.csv    NaN             0        SE Marsh     2015   
33   1082625.csv    NaN             0       KH Pandya     2017   
123  1175365.csv    NaN             0         PP Shaw     2019   
164  1178426.csv    NaN             0       JJ Bumrah     2019   

                            team                        team2 toss_decision  \
240        Kolkata Knight Riders             Rajasthan Royals         field   
304          Chennai Super Kings              Kings XI Punjab         field   
502          Sunrisers Hyderabad  Royal Challengers Bangalore           bat   
515  Royal Challengers Bangalore             Delhi Daredevils         field   
590        Kolkata Knight Riders             Rajasthan Royals           bat   
650             Rajasthan Royals              Kings XI Punjab         field   
33                 Gujarat Lions               Mumbai Indians           bat   
123               Delhi Capitals        Kolkata Knight Riders         field   
164               Mumbai Indians          Sunrisers Hyderabad           bat   

                     toss_winner                                   venue  \
240        Kolkata Knight Riders                                Newlands   
304          Chennai Super Kings                  "MA Chidambaram Stadiu   
502  Royal Challengers Bangalore      "Rajiv Gandhi International Stadiu   
515  Royal Challengers Bangalore                   M Chinnaswamy Stadium   
590             Rajasthan Royals                    Sheikh Zayed Stadium   
650              Kings XI Punjab                    "Sardar Patel Stadiu   
33                 Gujarat Lions  Saurashtra Cricket Association Stadium   
123               Delhi Capitals                        Feroz Shah Kotla   
164               Mumbai Indians                        Wankhede Stadium   

    winner winner_runs winner_wickets match_date  is_league_match  
240    tie         NaN            NaN 2009-04-23                0  
304    tie         NaN            NaN 2010-03-21                0  
502    tie         NaN            NaN 2013-04-07                0  
515    tie         NaN            NaN 2013-04-16                0  
590    tie         NaN            NaN 2014-04-29                0  
650    tie         NaN            NaN 2015-04-21                0  
33     tie         NaN            NaN 2017-04-29                0  
123    tie         NaN            NaN 2019-03-30                0  
164    tie         NaN            NaN 2019-05-02                0

In [21]:
for matchcsv in metadata.loc[np.logical_not(metadata.eliminator.isnull())].match_id.values:
    temp = pd.read_csv('balldata/'+matchcsv)
    print("Number of innings in {0} : {1}".format(matchcsv, temp.loc[:,'innings'].nunique()))

Number of innings in 392190.csv : 4
Number of innings in 419121.csv : 4
Number of innings in 598004.csv : 4
Number of innings in 598017.csv : 4
Number of innings in 729315.csv : 4
Number of innings in 829741.csv : 4
Number of innings in 1082625.csv : 4
Number of innings in 1175365.csv : 4
Number of innings in 1178426.csv : 4


<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
    <ul>
    <li>All the matches which have non-null value for <strong>eliminator</strong> represent the tie matches where the winner is decided by a super-over</li>
    <li>We can merge the <strong>winner</strong> column with <strong>eliminator</strong> column and have a new column <strong>'has_super_over'</strong> to show if the corresponding match led to a super-over ot not. </li>
    </ul>

In [22]:
#merging eliminator and winner attributes.

outcomes = []
for out, win in zip(metadata.eliminator.values, metadata.winner.values):
    if win=="tie":
        outcomes.append(out)
    else:
        outcomes.append(win)

metadata.loc[:, 'winner'] = outcomes
metadata['has_super_over'] = 1 - metadata.eliminator.isnull().astype(int)
metadata.drop('eliminator', axis=1, inplace=True)

In [23]:
metadata.has_super_over.value_counts()

0    742
1      9
Name: has_super_over, dtype: int64

<h2 style="text-align: center; font-size:2.5em;"> Method </h2>

In [24]:
print("Number of matches with 'method' attribute : {}".format(metadata.method.size - metadata.method.isnull().sum()))

Number of matches with 'method' attribute : 19


In [25]:
metadata.loc[np.logical_not(metadata.method.isnull())]

city            competition     match_id method  neutralvenue  \
214          Delhi  Indian Premier League   336022.csv    D/L             0   
217        Kolkata  Indian Premier League   336025.csv    D/L             0   
234      Cape Town  Indian Premier League   392183.csv    D/L             1   
237         Durban  Indian Premier League   392186.csv    D/L             1   
263      Centurion  Indian Premier League   392214.csv    D/L             1   
366          Kochi  Indian Premier League   501215.csv    D/L             0   
395        Kolkata  Indian Premier League   501245.csv    D/L             0   
405      Bangalore  Indian Premier League   501255.csv    D/L             0   
603          Delhi  Indian Premier League   733993.csv    D/L             0   
651  Visakhapatnam  Indian Premier League   829743.csv    D/L             0   
682      Hyderabad  Indian Premier League   829807.csv    D/L             0   
712      Hyderabad  Indian Premier League   980943.csv    D/L             0   
735        Kolkata  Indian Premier League   980989.csv    D/L             0   
739  Visakhapatnam  Indian Premier League   980997.csv    D/L             0   
740      Bangalore  Indian Premier League   980999.csv    D/L             0   
56       Bangalore  Indian Premier League  1082648.csv    D/L             0   
59          Jaipur  Indian Premier League  1136566.csv    D/L             0   
71         Kolkata  Indian Premier League  1136578.csv    D/L             0   
85           Delhi  Indian Premier League  1136592.csv    D/L             0   

      player_of_match   season                         team  \
214  DPMD Jayawardene  2007/08             Delhi Daredevils   
217           M Ntini  2007/08        Kolkata Knight Riders   
234        DL Vettori     2009             Delhi Daredevils   
237          CH Gayle     2009              Kings XI Punjab   
263         ML Hayden     2009          Chennai Super Kings   
366       BB McCullum     2011         Kochi Tuskers Kerala   
395     Iqbal Abdulla     2011        Kolkata Knight Riders   
405          CH Gayle     2011  Royal Challengers Bangalore   
603          DW Steyn     2014             Delhi Daredevils   
651         DA Warner     2015          Sunrisers Hyderabad   
682           V Kohli     2015          Sunrisers Hyderabad   
712          AB Dinda     2016          Sunrisers Hyderabad   
735         YK Pathan     2016        Kolkata Knight Riders   
739          AB Dinda     2016      Rising Pune Supergiants   
740           V Kohli     2016  Royal Challengers Bangalore   
56    NM Coulter-Nile     2017          Sunrisers Hyderabad   
59          SV Samson     2018             Rajasthan Royals   
71           KL Rahul     2018        Kolkata Knight Riders   
85            RR Pant     2018             Delhi Daredevils   

                           team2 toss_decision                  toss_winner  \
214              Kings XI Punjab           bat             Delhi Daredevils   
217          Chennai Super Kings           bat        Kolkata Knight Riders   
234              Kings XI Punjab         field             Delhi Daredevils   
237        Kolkata Knight Riders         field        Kolkata Knight Riders   
263              Kings XI Punjab           bat          Chennai Super Kings   
366          Chennai Super Kings         field         Kochi Tuskers Kerala   
395          Chennai Super Kings           bat          Chennai Super Kings   
405        Kolkata Knight Riders         field  Royal Challengers Bangalore   
603          Sunrisers Hyderabad         field          Sunrisers Hyderabad   
651        Kolkata Knight Riders         field        Kolkata Knight Riders   
682  Royal Challengers Bangalore           bat          Sunrisers Hyderabad   
712      Rising Pune Supergiants         field      Rising Pune Supergiants   
735      Rising Pune Supergiants           bat      Rising Pune Supergiants   
739             Delhi Daredevils         field      Rising Pune Su

In [26]:
metadata.method.value_counts()

D/L    19
Name: method, dtype: int64

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
<ul>
    <li>All the matches, where the final outcome is decided by <a href="https://en.wikipedia.org/wiki/Duckworth%E2%80%93Lewis%E2%80%93Stern_method" target="_blank"><b>DLS</b> method</a> has <b>method</b> attribute set to the value <b>"D/L"</b></li>
    <li>DLS method is only used, if rain(or any other disturbance) has delayed the match for significant time.</li>
    <li> Hence, <b>method</b> attribute can be converted into a <b>boolean holder</b></li>
</ul>
</div>

In [27]:
metadata.loc[:, 'method'] = np.logical_not(meta.method.isnull()).astype(int)
metadata.method.value_counts()

0    732
1     19
Name: method, dtype: int64

<h2 style="text-align: center; font-size:2.5em;"> Player_of_match </h2>

In [28]:
metadata.loc[metadata.player_of_match.isnull()]

city            competition     match_id  method  neutralvenue  \
415      Delhi  Indian Premier League   501265.csv       0             0   
660  Bangalore  Indian Premier League   829763.csv       0             0   
685  Bangalore  Indian Premier League   829813.csv       0             0   
162  Bengaluru  Indian Premier League  1178424.csv       0             0   

    player_of_match season                         team             team2  \
415             NaN   2011             Delhi Daredevils     Pune Warriors   
660             NaN   2015  Royal Challengers Bangalore  Rajasthan Royals   
685             NaN   2015  Royal Challengers Bangalore  Delhi Daredevils   
162             NaN   2019  Royal Challengers Bangalore  Rajasthan Royals   

    toss_decision                  toss_winner                  venue  \
415           bat             Delhi Daredevils       Feroz Shah Kotla   
660         field             Rajasthan Royals  M Chinnaswamy Stadium   
685         field  Royal Challengers Bangalore  M Chinnaswamy Stadium   
162         field             Rajasthan Royals  M.Chinnaswamy Stadium   

        winner winner_runs winner_wickets match_date  is_league_match  \
415  no result         NaN            NaN 2011-05-21                0   
660  no result         NaN            NaN 2015-04-29                0   
685  no result         NaN            NaN 2015-05-17                0   
162  no result         NaN            NaN 2019-04-30                0   

     has_super_over  
415               0  
660               0  
685               0  
162               0

In [29]:
metadata.winner.value_counts()['no result']

4

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
<ul>
    <li>There are 4 matches with no 'player_of_match' attribute.</li>
    <li>Further inspection shows all these four matches have attribute <b>method</b> set to <b>"D/L"</b>.</li>
    <li>Rain wiped out the whole time allocated to the corresponding match. Hence we neither have a winner nor a player_of_match for such matches.</li>
</ul>
</div>

In [30]:
metadata.isnull().sum()

city                 0
competition          0
match_id             0
method               0
neutralvenue         0
player_of_match      4
season               0
team                 0
team2                0
toss_decision        0
toss_winner          0
venue                0
winner               0
winner_runs        414
winner_wickets     350
match_date           0
is_league_match      0
has_super_over       0
dtype: int64

<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
<div>Our data looks much clean now :)</div>
</div>

<h2 style="text-align: center; font-size:2.5em;"> Feature Augmentation </h2>
<div style="border: 1px solid black; padding: 1.5em; line-height:2; border-radius: 15px;">
    <div>This is the time to reinforce our analysis with some custom variables<div>
    <div>
        <div>New variables to add in our data:</div>
        <ul>
            <li>toss_looser :- Team which lost the toss.</li>
            <li>bat1 :- first batting team</li>
            <li>bat2 :- second batting team</li>
            <li>toss_winner_is_match_winner :- <b>True</b> if the team which won the <b>toss</b> eventually <b>wins the match</b></li>
            <li>score1 :- Runs scored by the team batting first.</li>
            <li>score2 :- Runs scored bt the team batting second.</li>
            <li>runrate_diff :- difference in runrate of both the teams</li>
            <li>balls1 :- Balls faced by team batting first</li>
            <li>balls2 :- Balls faced by team batting second</li>
            <li>mom_team :- Team of "player of match"</li>
        </ul>
    </div>
    <div>
        <div>But, before moving ahead let's make some changes to the structure of the data.</div>
        <ul>
            <li>Rename <b>"team"</b> and <b>"team2"</b> as <b>"home_team"</b> and <b>"away_team"</b> respectively</li>
            <li>Create a LabelEncoder mapping team name to an integer</li>
            <li>Encode team names according to the above created LabelEncoder in the whole dataset.</li>
            <li>Remove the redundant <b>competition</b> variable</li>
        </ul>
    </div>
</div>

In [31]:
metadata.rename({'team':'home_team', 'team2':'away_team'}, axis=1, inplace=True)

In [32]:
metadata.drop('competition', axis=1, inplace=True)

In [33]:
team_name_encoder = LabelEncoder()
team_name_encoder.fit(metadata.winner.values)
team_name_encoder.classes_.size

16

In [34]:
for idx,team in enumerate(team_name_encoder.classes_):
    print('{}. {}'.format(idx+1, team))

1. Chennai Super Kings
2. Deccan Chargers
3. Delhi Capitals
4. Delhi Daredevils
5. Gujarat Lions
6. Kings XI Punjab
7. Kochi Tuskers Kerala
8. Kolkata Knight Riders
9. Mumbai Indians
10. Pune Warriors
11. Rajasthan Royals
12. Rising Pune Supergiant
13. Rising Pune Supergiants
14. Royal Challengers Bangalore
15. Sunrisers Hyderabad
16. no result


In [35]:
metadata.replace({"Delhi Capitals": "Delhi Daredevils", "Rising Pune Supergiant": "Rising Pune Supergiants"}, inplace=True)

In [36]:
team_name_encoder = LabelEncoder()
team_name_encoder.fit(metadata.winner.values)
team_name_encoder.classes_.size

14

In [37]:
#encoding "team name"
#for col in ['home_team', 'away_team', 'toss_winner', 'winner']:
#   metadata[col] = team_name_encoder.transform(metadata[col])

In [38]:
loosers = []
for home, away, winner in zip(metadata.home_team.values, metadata.away_team.values, metadata.toss_winner.values):
    if winner == home:
        loosers.append(away)
    else:
        loosers.append(home)
        
metadata['toss_looser'] = loosers

In [39]:
bat1=[]
bat2=[]

for decision, winner, looser in zip(metadata.toss_decision.values, metadata.toss_winner.values, metadata.toss_looser.values):
    if decision == "bat":
        bat1.append(winner)
        bat2.append(looser)
    else:
        bat1.append(looser)
        bat2.append(winner)

metadata['bat1'] = bat1
metadata['bat2'] = bat2

In [40]:
metadata['toss_winner_is_winner'] = (metadata['toss_winner']==metadata['winner'])
metadata['toss_winner_is_winner'] = metadata['toss_winner_is_winner'].astype(int)

In [41]:
def score(match, innings):
    delivery_data = pd.read_csv('balldata/'+match)
    if innings > delivery_data['innings'].nunique():
        return np.nan
    
    delivery_data = delivery_data.groupby('innings').sum().loc[innings]
    return delivery_data['runs']+delivery_data['extras']
    
def extras(match, innings):
    delivery_data = pd.read_csv('balldata/'+match)
    if innings> delivery_data['innings'].nunique():
        return np.nan
    return delivery_data.groupby('innings').sum()['extras'][innings]

def delivery_count(match, innings):
    delivery_data = pd.read_csv('balldata/'+match)
    if innings > delivery_data['innings'].nunique():
        return np.nan
    return delivery_data.loc[delivery_data.innings==innings].shape[0]

def runrate(match, innings):
    runs = score(match, innings)
    delivery_data = pd.read_csv('balldata/'+match)
    if innings > delivery_data['innings'].nunique():
        return None
    
    lastball = delivery_data.loc[delivery_data.innings==innings]['over'].values[-1]
    legal_balls = (lastball//1)*6 + (lastball*10)%10
    
    return (6*runs)/legal_balls

In [42]:
start = time.time()
score1 = []
score2 = []
extras1 = []
extras2 = []
balls1 = []
balls2 =[]
runrate1 = []
runrate2 = []

for match in metadata.match_id.values:
    score1.append(score(match, 1))
    extras1.append(extras(match, 2))
    balls1.append(delivery_count(match, 1))
    runrate1.append(runrate(match, 1))
    
    score2.append(score(match, 2))
    extras2.append(extras(match, 1))
    balls2.append(delivery_count(match, 2))
    runrate2.append(runrate(match, 2))

metadata['score1'] = score1
metadata['score2'] = score2
metadata['extras1'] = extras1
metadata['extras2'] = extras2
metadata['balls1'] = balls1
metadata['balls2'] = balls2
metadata['runrate1'] = runrate1
metadata['runrate2'] = runrate2

print("Excecution Time : {0:.3f} s".format(time.time()-start))

Excecution Time : 26.221 s


In [43]:
metadata.head()

city    match_id  method  neutralvenue player_of_match   season  \
174   Bangalore  335982.csv       0             0     BB McCullum  2007/08   
176       Delhi  335984.csv       0             0     MF Maharoof  2007/08   
175  Chandigarh  335983.csv       0             0      MEK Hussey  2007/08   
177      Mumbai  335985.csv       0             0      MV Boucher  2007/08   
178     Kolkata  335986.csv       0             0       DJ Hussey  2007/08   

                       home_team                    away_team toss_decision  \
174  Royal Challengers Bangalore        Kolkata Knight Riders         field   
176             Delhi Daredevils             Rajasthan Royals           bat   
175              Kings XI Punjab          Chennai Super Kings           bat   
177               Mumbai Indians  Royal Challengers Bangalore           bat   
178        Kolkata Knight Riders              Deccan Chargers           bat   

                     toss_winner                               venue  \
174  Royal Challengers Bangalore               M Chinnaswamy Stadium   
176             Rajasthan Royals                    Feroz Shah Kotla   
175          Chennai Super Kings  "Punjab Cricket Association Stadiu   
177               Mumbai Indians                    Wankhede Stadium   
178              Deccan Chargers                        Eden Gardens   

                          winner winner_runs winner_wickets match_date  \
174        Kolkata Knight Riders         140            NaN 2008-04-18   
176             Delhi Daredevils         NaN              9 2008-04-19   
175          Chennai Super Kings          33            NaN 2008-04-19   
177  Royal Challengers Bangalore         NaN              5 2008-04-20   
178        Kolkata Knight Riders         NaN              5 2008-04-20   

     is_league_match  has_super_over                  toss_looser  \
174                0               0        Kolkata Knight Riders   
176                0               0             Delhi Daredevils   
175                0               0              Kings XI Punjab   
177                0               0  Royal Challengers Bangalore   
178                0               0        Kolkata Knight Riders   

                      bat1                         bat2  \
174  Kolkata Knight Riders  Royal Challengers Bangalore   
176       Rajasthan Royals             Delhi Daredevils   
175    Chennai Super Kings              Kings XI Punjab   
177         Mumbai Indians  Royal Challengers Bangalore   
178        Deccan Chargers        Kolkata Knight Riders   

     toss_winner_is_winner  score1  score2  extras1  extras2  balls1  balls2  \
174                      0   222.0    82.0     19.0       17     124   101.0   
176                      0   129.0   132.0     10.0        7     122    97.0   
175                      1   240.0   207.0     11.0        6     124   124.0   
177                      0   165.0   166.0      5.0       11     123   123.0   
178                      0   110.0   112.0     28.0       10     118   122.0   

      runrate1   runrate2  
174  11.100000   5.347826  
176   6.450000   8.703297  
175  12.000000  10.350000  
177   8.250000   8.369748  
178   5.892857   5.894737

In [44]:
metadata.index = np.arange(metadata.shape[0]).astype(int)

In [45]:
metadata.head()

city    match_id  method  neutralvenue player_of_match   season  \
0   Bangalore  335982.csv       0             0     BB McCullum  2007/08   
1       Delhi  335984.csv       0             0     MF Maharoof  2007/08   
2  Chandigarh  335983.csv       0             0      MEK Hussey  2007/08   
3      Mumbai  335985.csv       0             0      MV Boucher  2007/08   
4     Kolkata  335986.csv       0             0       DJ Hussey  2007/08   

                     home_team                    away_team toss_decision  \
0  Royal Challengers Bangalore        Kolkata Knight Riders         field   
1             Delhi Daredevils             Rajasthan Royals           bat   
2              Kings XI Punjab          Chennai Super Kings           bat   
3               Mumbai Indians  Royal Challengers Bangalore           bat   
4        Kolkata Knight Riders              Deccan Chargers           bat   

                   toss_winner                               venue  \
0  Royal Challengers Bangalore               M Chinnaswamy Stadium   
1             Rajasthan Royals                    Feroz Shah Kotla   
2          Chennai Super Kings  "Punjab Cricket Association Stadiu   
3               Mumbai Indians                    Wankhede Stadium   
4              Deccan Chargers                        Eden Gardens   

                        winner winner_runs winner_wickets match_date  \
0        Kolkata Knight Riders         140            NaN 2008-04-18   
1             Delhi Daredevils         NaN              9 2008-04-19   
2          Chennai Super Kings          33            NaN 2008-04-19   
3  Royal Challengers Bangalore         NaN              5 2008-04-20   
4        Kolkata Knight Riders         NaN              5 2008-04-20   

   is_league_match  has_super_over                  toss_looser  \
0                0               0        Kolkata Knight Riders   
1                0               0             Delhi Daredevils   
2                0               0              Kings XI Punjab   
3                0               0  Royal Challengers Bangalore   
4                0               0        Kolkata Knight Riders   

                    bat1                         bat2  toss_winner_is_winner  \
0  Kolkata Knight Riders  Royal Challengers Bangalore                      0   
1       Rajasthan Royals             Delhi Daredevils                      0   
2    Chennai Super Kings              Kings XI Punjab                      1   
3         Mumbai Indians  Royal Challengers Bangalore                      0   
4        Deccan Chargers        Kolkata Knight Riders                      0   

   score1  score2  extras1  extras2  balls1  balls2   runrate1   runrate2  
0   222.0    82.0     19.0       17     124   101.0  11.100000   5.347826  
1   129.0   132.0     10.0        7     122    97.0   6.450000   8.703297  
2   240.0   207.0     11.0        6     124   124.0  12.000000  10.350000  
3   165.0   166.0      5.0       11     123   123.0   8.250000   8.369748  
4   110.0   112.0     28.0       10     118   122.0   5.892857   5.894737

In [46]:
print("notebook excecution time : ", end="")
print("{:.3f} s".format(time.time() - notebook_start_time))

notebook excecution time : 51.602 s


In [47]:
metadata.to_csv('metadata/metadata.csv', index=False)